db_service/dao_generator.go

```diff
@@ -0,0 +1,342 @@
+	record := models.{{.Type}}{}
+	if err := ds.db.Unscoped().Where("{{.PrimaryKeyField}} = ?", pk).First(&record).Error; err != nil {
+		return false, err
+	}
+
+	return record.DeletedAt != nil, nil
+}
+
+{{- end }}
+`))
+
+var daoTestTemplate = template.Must(template.New("").Funcs(
+	template.FuncMap{
+		"funcName": functionNameGen,
+	},
+).Parse(`// Copyright {{ .Timestamp.Year }} the Service Broker Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated by go generate; DO NOT EDIT.
+// This file was generated at {{ .Timestamp }}.
+
+package db_service
+
+import (
+	"testing"
+	"time"
+
+  "github.com/GoogleCloudPlatform/gcp-service-broker/brokerapi/brokers/models"
+	"github.com/jinzhu/gorm"
+)
+
+func newInMemoryDatastore(t *testing.T) *SqlDatastore {
+	testDb, err := gorm.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Fatalf("Error opening test database %s", err)
+	}
+
+	{{range .Models}}testDb.CreateTable(models.{{.Type}}{})
+	{{end}}
+	return &SqlDatastore{db: testDb}
+}
+
+
+{{- range .Models}}
+
+func TestSqlDatastore_{{.Type}}DAO(t *testing.T) {
+	ds := newInMemoryDatastore(t)
+	testPk := {{.PrimaryKeyType}}(42)
+
+	instance := models.{{.Type}}{}
+	instance.ID = testPk
+
+	// on startup, there should be no objects to find or delete
+	if count, err := ds.{{funcName "Count" . .PrimaryKeyField}}(testPk); count != 0 || err != nil {
+		t.Fatalf("Expected count to be 0 and error to be nil got count: %d, err: %v", count, err)
+	}
+
+	if _, err := ds.{{funcName "Get" . .PrimaryKeyField}}(testPk); err != gorm.ErrRecordNotFound {
+		t.Errorf("Expected an ErrRecordNotFound trying to get non-existing PK got %v", err)
+	}
+
+	if _, err := ds.{{funcName "CheckDeleted" . .PrimaryKeyField}}(testPk); err != gorm.ErrRecordNotFound {
+		t.Errorf("Expected an ErrRecordNotFound trying to check deletion status of a non-existing PK got %v", err)
+	}
+
+	if err := ds.{{funcName "Delete" . .PrimaryKeyField}}(testPk); err != gorm.ErrRecordNotFound {
+		t.Errorf("Expected an ErrRecordNotFound trying to delete non-existing PK got %v", err)
+	}
+
+	// Should be able to create the item
+	beforeCreation := time.Now()
+	if err := ds.{{funcName "Create" .}}(&instance); err != nil {
+		t.Errorf("Expected to be able to create the item %#v, got error: %s", instance, err)
+	}
+	afterCreation := time.Now()
+
+	// after creation we should be able to get the item
+	ret, err := ds.{{funcName "Get" . .PrimaryKeyField}}(testPk)
```

Comment:
Consider adding a test step which compares the equality of the retrieved object verses the original.

Reply:
It looks like the timezones are changed between serialization/deserialization even if the times (I _hate_ dates) are the same so I can't `.DeepEqual`, what do you think about me adding a `map[string]interface{}` to the generator for field/example pairs and do a field-by-field set and comparison from those? It feels kinda wonky because now the examples won't be type checked up until `go generate` is run but it feels better than reflection in templates.

Reply:
Is gorm messing with `time.Time` fields on deserialization?

Reply:
The times do represent the same point in time, but aren't created/deserialized in the same way so they can't be tested using reflection.

If I inject this line here's the output:

```
// t.Errorf("test %#v %#v, %#v", instance.CreatedAt.Equal(ret.CreatedAt), instance.CreatedAt, ret.CreatedAt)

dao_test.go:79: test true time.Time{wall:0xbed77c4956a625a3, ext:21705095, loc:(*time.Location)(0xfb4740)}, time.Time{wall:0x16a625a3, ext:63670546597, loc:(*time.Location)(0xc42046a600)}
```

The strings are something like the following: `2018-08-22 07:54:10.142900966 -0700 PDT m=+0.009016746` vs `2018-08-22 07:54:10.142900966 -0700 -0700`.

So I think the best courses of action are to either defer testing the serialization/deserialization to an integration test when it's used by another module or specify values field-by-field and test those explicitly which would keep it part of the code generation so it wouldn't be forgotten in future additions but still be flexible-ish enough.

Reply:
A field by field test should work. One option to potentially make the code less verbose and more general is to have the models maintain a list of fields to exclude from the equality test. Here's an example of using reflect which could potentially be useful in this scenario: https://gist.github.com/drewolson/4771479
YMMV :)